src/bookshelf/src/bksw_sec.erl

```diff
@@ -35,109 +37,283 @@ is_authorized(Req0, #context{} = Context) ->
-        {StringToSign, Signature} ->
-            case ExpireDiff =< 0 of
-                true ->
-                    ?LOG_DEBUG("req_id=~p expired signature (~p) for ~p",
-                               [ReqId, Expires, Path]),
-                    encode_access_denied_error_response(RequestId, Req0, Context);
-                false ->
-                    case ((erlang:iolist_to_binary(AWSAccessKeyId) ==
-                               erlang:iolist_to_binary(AccessKey)) andalso
-                          erlang:iolist_to_binary(Signature) ==
-                              erlang:iolist_to_binary(IncomingSignature)) of
-                        true ->
-                            MaxAge = "max-age=" ++ integer_to_list(ExpireDiff),
-                            Req1 = wrq:set_resp_header("Cache-Control", MaxAge, Req0),
-                            {true, Req1, Context};
-                        false ->
-                            ?LOG_DEBUG("req_id=~p signing error for ~p", [ReqId, Path]),
-                            encode_sign_error_response(AWSAccessKeyId, IncomingSignature, RequestId,
-                                                       StringToSign, Req0, Context)
-                    end
-                end;
-        error ->
-            encode_access_denied_error_response(RequestId, Req0, Context)
+% presigned url verification
+% https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html
+presigned_auth(RequestId, Req0, Context, Headers0) ->
+    Credential = wrq:get_qs_value("X-Amz-Credential", "", Req0),
+    XAmzDate = wrq:get_qs_value("X-Amz-Date", "", Req0),
+    SignedHeaderKeysString = wrq:get_qs_value("X-Amz-SignedHeaders", "", Req0),
+    IncomingSignature = wrq:get_qs_value("X-Amz-Signature", "", Req0),
+    % 1 =< XAmzExpires =< 604800
+    XAmzExpiresString = wrq:get_qs_value("X-Amz-Expires", "", Req0),
+    common_auth(RequestId, Req0, Context, Credential, XAmzDate, SignedHeaderKeysString, IncomingSignature, XAmzExpiresString, Headers0, presigned_url).
+
+% https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-auth-using-authorization-header.html
+header_auth(RequestId, IncomingAuth, Req0, Context, Headers0) ->
+    try
+        (ParseAuth = parse_authorization(IncomingAuth)) /= err orelse throw({RequestId, Req0, Context}),
+        [Credential, SignedHeaderKeysString, IncomingSignature] = ParseAuth,
+        XAmzDate = wrq:get_req_header("x-amz-date", Req0),
+        common_auth(RequestId, Req0, Context, Credential, XAmzDate, SignedHeaderKeysString, IncomingSignature, "300", Headers0, authorization_header)
+    catch
+        _ -> encode_access_denied_error_response(RequestId, Req0, Context)
     end.
 
-make_signed_url_authorization(SecretKey, Method, Path, Expires, Headers) ->
+common_auth(RequestId, Req0, #context{reqid = ReqId} = Context, Credential, XAmzDate, SignedHeaderKeysString, IncomingSignature, XAmzExpiresString, Headers0, VerificationType) ->
```

Comment:
It is somewhat long.  You mean break it up into chunks perhaps?

Reply:
I don't worry about absolute length, functions can be as long as they need to be.  I also aren't as deep into this, so perhaps these suggestions are off base:

1. Either let's try to avoid the throws _or_ clearly document what gets thrown and why in a comment above the function head.  If the throws are the best way to handle errors here, then let's also organize the calling code in a way that it makes it very clear where the throws from _this_ function and only this function are handled.

2. Can we somehow split out the bits of this that are truly common from the bits that are specific to particular verification types?

3. You've added some nice type annotations to other functions.  Are there bits of this we can split out into well-typed functions with descriptive names?  We shouldn't go overboard so that we are constantly jumping between functions when reading the code, but my guess is there are one or two functions that could be extracted here.  

I'm out today, but if you are interested I'm happy to pair a bit on this since it might also just be my lack of context here.

Reply:
Addressing #2 above, here are some notes on the common_auth function to help with your review.  The top part of the function sets up a bunch of variables that are used by both verification types.  Pursuant to this, I added some temporary comments in the code, each labeled 'CODE REVIEW,' for your viewing pleasure.

Then the function splits into two paths - the presigned_url verification path, and the authorization_header path - which handle processing particular to those verification types, although as you can see, both parts have the same structure.  This 'split path' is handled by the case statement in the middle of the function.

The paths then join again into processing common to both - comparison of incoming vs. computed signatures.

At least upon casual or cursory examination, the only significant part of the function I can see that can reasonably be split-out is the innards of the case statement (the 'split path').  Would you like me to factor this part out?  It doesn't represent processing common to both, so in a way it arguably doesn't belong in a function called 'common_auth.'  Also, does anything else jump out at you that can be factored-out, i.e. have I missed anything?

Here's a crude, simplistic diagram showing the structure explained above.  The 'diamond' in the middle is where I'm seeing the potential for refactor, assuming you think it's worthwhile to do that.  By all means let me know if there's anything else.

![image](https://user-images.githubusercontent.com/51833247/89885083-1e660f00-db90-11ea-9533-c819122e3ce1.png)